src/StackExchange.Utils.Configuration/SubstitutionHelper.cs

```diff
@@ -0,0 +1,169 @@
+            var matches = _substitutionPattern.Matches(value);
+            if (matches.Count > 0)
+            {
+                // detect cycles - this can happen when there are multiple
+                // substitutions that reference values in other child providers.
+                // This will end up overflowing the stack so we detect it here
+                // and throw an exception
+                var cycleDetector = _cycleDetector.Value ??= cycleDetectorFactory();
+                
+                cycleDetector.Add(key);
+                
+                // we have some things to substitute, so manipulate the value as a span
+                // so we can slice and dice the substitution values
+                var valueBuilder = new StringBuilder(value.Length);
+                var valueSpan = value.AsSpan();
+                var index = 0;
+
+                try
+                {
+                    foreach (Match match in matches)
+                    {
+                        var substitutionKey = match.Groups["key"].Value;
+                        if (cycleDetector.Contains(substitutionKey))
+                        {
+                            ThrowCycleDetectedError(key, substitutionKey);
+                        }
+
+                        var substitutionValue = substitutions[substitutionKey];
+                        if (substitutionValue != null)
+                        {
+                            // rewrite the value with the substitution
+                            valueBuilder.Append(
+                                valueSpan.Slice(index, match.Index - index)
+                            );
+                            
+                            valueBuilder.Append(substitutionValue);
+                            index = match.Index + match.Length;
+                        }
+                    }
+                }
+                finally
+                {
+                    cycleDetector.Remove(key);
+                }
+                
+                // copy across the remainder of the string
+                valueBuilder.Append(
+                    valueSpan.Slice(index, value.Length - index)
+                );
+
+                value = valueBuilder.ToString();
+            }
+
+            return value;
+        }
+        
+        private static void ThrowCycleDetectedError(string key, string substitutionKey) => throw new InvalidOperationException(
+            $"Cycle detected for when trying to substitute [{substitutionKey}] in [{key}]"
+        );
+        
+        private class CycleDetector : ICycleDetector
+        {
+            private readonly HashSet<string> _values = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            public void Add(string key) => _values.Add(key);
+
+            public void Remove(string key) => _values.Remove(key);
+
+            public bool Contains(string key) => _values.Contains(key);
+        }
```

Comment:
This works, but FWIW could simplify a bit with a `Stack<string>` in the `AsyncLocal` as well I think.

Reply:
Might be missing something here, but I'm not sure a `Stack<string>` works here? We're trying to detect cycles so if:

`keyA = ${keyB}`
`keyB = ${keyC}`
`keyC = ${keyA}`

We'd need to go enumerate all values in the stack? Only reason this is abstracted out to an interface is that we can't really do cycle detection for the simple value (not provider-based) replacement case so there's a no-op implementation